aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -514,6 +514,141 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+//    Example:
+//      1 1 0
+//      1 1 1
+//      1 1 1
+//    In this case, we first calculate the size of top trapezoid, and then
+//    calculate the size of the bottom rectangle.
+int64_t get_tril_size(int64_t row, int64_t col, int64_t offset) {
+  // number of elements in the first row of the tril
+  auto m_first_row = offset > 0 ?
+    std::min<int64_t>(col, 1 + offset) : // upper bounded by col
+    row + offset > 0; // either 0 or 1
+  // number of elements in the last row of the tril, bounded by [0, col]
+  auto m_last_row = std::max<int64_t>(0, std::min<int64_t>(col, row + offset));
+  // number of rows, bounded by [0, row]
+  auto n_row_all = std::max<int64_t>(0, std::min<int64_t>(row, row + offset));
+  auto n_row_trapezoid = (m_last_row - m_first_row + 1);
+
+  // calculate # of elements in the top trapezoid
+  auto n_indices =
+    (m_first_row + m_last_row) * n_row_trapezoid >> 1;
+
+  // calculate # of elements in the bottom rectangle if there is any
+  auto diff_row = n_row_all - n_row_trapezoid;
+  if (diff_row > 0) {
+    n_indices += diff_row * col;
+  }
+
+  return n_indices;
+}
+} // namespace
+
+Tensor tril_indices(
+    int64_t row, int64_t col, int64_t offset, const TensorOptions& options) {
+  AT_CHECK(row >= 0, "row must be non-negative, got", row);
+  AT_CHECK(col >= 0, "col must be non-negative, got", col);
+
+  auto n_indices = get_tril_size(row, col, offset);
+
+  // create an empty Tensor with correct soize
+  auto result = at::empty({n_indices, 2}, options);
+
+  // The following three design options result in very little performance
+  // differences. Hence, the 3rd option is taken for simpler code. Refer to
+  // #14904 for more details.
+  //
+  // 1. sequential RAM access: fill row coordinates first, then columns. This
+  //    results in two for-loop and more arithmetic operations.
+  //
+  // 2. interleaved RAM access: fill in index coordinates one by one, which
+  //    jumps between the two output Tensor rows in every iteration.
+  //
+  // 3. sequential RAM + transpose: create a n X 2 Tensor, fill the Tensor
+  //    sequentially, and then tranpose it.
+  AT_DISPATCH_ALL_TYPES(result.type(), "tril_indices", [&]() -> void {
+    // fill the Tensor with correct values
+    scalar_t* result_data = result.data<scalar_t>();
+    int64_t i = 0;
+
+    scalar_t r = std::max<int64_t>(0, -offset), c = 0;
+    n_indices = n_indices << 1;
+    while (i < n_indices) {
+      result_data[i++] = r;
+      result_data[i++] = c;
+
+      // move to the next column and check if (r, c) is still in bound
+      c += 1;
+      if (c > r + offset || c >= col) {
+        r += 1;
+        c = 0;
+        // NOTE: not necessary to check if r is less than row here, because i
+        // and n_indices provide the guarantee
+      }
+    }
+  });
+
+  return result.transpose(0, 1);
+}
+
+
+Tensor triu_indices(
+    int64_t row, int64_t col, int64_t offset, const TensorOptions& options) {
```

Comment:
Because we accept TensorOptions& (this is a side effect of this function being a tensor creation method), I think the function in python-land will take the following arguments: `device, requires_grad, layout`. Could you check to see if the function does take these in python land and make sure this code throws an error if anything incorrect is passed in, and then write some tests for this?
The function currently supports `device='cpu'`, `layout=torch.strided`, `requires_grad=True`.

Reply:
Yes, python-land indeed takes `device, requires_grad, layout`.  Tests are added to make sure we only accept `device=cpu`, and `layout=torch.strided`.